File: src/rb/uptime.rb
#!/usr/bin/env ruby

require 'ffi'

current_os = Gem::Platform.local.os

if current_os == 'darwin'
  module Sysctl
    extend FFI::Library
    ffi_lib FFI::Library::LIBC
    attach_function :sysctlbyname, %i[string pointer pointer pointer ulong], :int

    class Timeval < FFI::Struct
      layout :tv_sec, :long,
             :tv_usec, :int32
    end
  end

  len = FFI::MemoryPointer.new(:uint)
  len.write_uint(Sysctl::Timeval.size)
  boottime = Sysctl::Timeval.new
  Sysctl.sysctlbyname('kern.boottime', boottime, len, nil, 0)
  puts Time.now.to_i - boottime[:tv_sec]

elsif current_os == 'linux'
  module Sys
    extend FFI::Library
    ffi_lib FFI::Library::LIBC

    class Sysinfo < FFI::Struct
      layout :uptime, :long,
             :loads, [:ulong, 3],
             :totalram, :ulong,
             :freeram, :ulong,
             :sharedram, :ulong,
             :bufferram, :ulong,
             :totalswap, :ulong,
             :freeswap, :ulong,
             :procs, :ushort,
             :pad, :ushort,
             :totalhigh, :ulong,
             :freehigh, :ulong,
             :mem_unit, :int,
             :_f, [:char, 0] # padding
    end

    attach_function :sysinfo, [Sysinfo.by_ref], :int
  end

  info = Sys::Sysinfo.new
  Sys.sysinfo(info)

  puts info[:uptime] # system uptime in seconds
end

File: src/c/uptime.c
#include <stdio.h>

#if defined(__APPLE__) || defined(__FreeBSD__)
#include <stdlib.h>
#include <sys/sysctl.h>
#include <sys/time.h>
#include <sys/types.h>
#include <time.h>
#include <unistd.h>

#elif defined(__linux__)
#include <sys/sysinfo.h>
#endif

int main(void) {
#if defined(__APPLE__) || defined(__FreeBSD__)
  struct timeval boottime;
  size_t len = sizeof(boottime);
  int mib[2] = {CTL_KERN, KERN_BOOTTIME};  // MIB for kern.boottime

  // Retrieve the boottime
  if (sysctl(mib, 2, &boottime, &len, NULL, 0) < 0) {
    perror("sysctl");
    exit(EXIT_FAILURE);
  }

  time_t now;
  time(&now);                                     // Get the current time
  time_t uptime_seconds = now - boottime.tv_sec;  // Calculate uptime in seconds

  printf("%ld\n", uptime_seconds);
  return EXIT_SUCCESS;
#endif

#if defined(__linux__)
  struct sysinfo info;
  sysinfo(&info);
  printf("%ld\n", info.uptime);
#endif
}
